# Analyzing MiSeq and iSeq fastq files using DADA2
Both Illumina MiSeq and iSeq can generate sequences that are suitable for 16S rRNA amplicon sequence analysis.
However, fastq files that are generated by iSeq contain only three types of Q-scores, 11 (,), 25 (;) and 37 (F). In this folder, I briefly examined how these differences influence DADA2 sequence analysis (<a href="https://doi.org/10.1038/nmeth.3869">Callahan et al. 2016 Nature Methods</a>).

<!--
According to the analysis, MiSeq-style and iSeq-style fastq files generated very similar results. Both fastq files seem to be appropriately analyzed by DADA2.
-->

# General procedure
0. Partial 16S rRNA sequences were amplified using 515F-806R, and the amplicons were sequenced by MiSeq V2 250 x 2 bp kit.
1. Started from MiSeq fastq files (0-39) (Example fastq files are in 02_DemultiplexedOut_MiSeqStyle). Note that, although I analyzed 252 fastq files (= 126 samples), only six fastq files are in 02_DemultiplexedOut_MiSeqStyle because the results have not been published yet (a preprint is available <a href="https://doi.org/10.1101/2020.04.08.032524">here</a>).
2. Manually converted MiSeq Q-scores to iSeq Q-scores using a shell script (Example fastq files are in 02_DemultiplexedOut_iSeqStyle).
3. These two types of fastq files were analyzed identically using DADA2 ("03_SeqAnalysisDADA2_MiSeq.R" and "03_SeqAnalysisDADA2_iSeq.R").
4. Representative sequences were saved as "ASV.fa", and taxa information was assigned using Claident (<a href="https://doi.org/10.1371/journal.pone.0076910">Tanabe & Toju 2013 PLoS ONE</a>).
5. ASV table, sample information and taxa information were imported as phyloseq object.
6. Three types of visualizations were done: Barplots of MiSeq-style and iSeq-style fastq files (__Figure 1__), sequence reads of MiSeq-style v.s. iSeq-style fastq files (__Figure 2__) and relative abundance of MiSeq-style v.s. iSeq-style fastq files (__Figure 3__).

### Notes
- I converted 250 x 2 bp MiSeq fastq files to iSeq-style fastq files, but the maximum length of iSeq sequencer is currently 150 x 2 bp.
- I used `bcl2fastq` command to generate (MiSeq-style) fastq files from BCL files, and demultiplexed them using `clsplitseq` function implemented in Claident. This procedure removed the forward primer region when demultiplesing was performed.
- In the library preparation process, I included artifitial prokaryote-like sequences as internal standard DNAs in order to quantify prokaryotic DNAs. There internal standard DNAs occupy a large proportion of total sequence reads (ca. 35%). For details, see <a href="https://doi.org/10.1101/2020.04.08.032524">Ushio (2020) <i>bioRxiv</i></a>.

## Figure 1: Barplots of MiSeq-style v.s. iSeq-style fastq files
<img src="06_ComparisonOut/MiSeq_vs_iSeq_barplot.jpg" width="1000px">
The upper and lower panels indicate barplots of MiSeq-style and iSeq-style fastq files, respectively. Both show very similar patterns.

## Figure 2: Sequence reads of MiSeq-style v.s. iSeq-style fastq files
<img src="06_ComparisonOut/SequenceReads_scatterplot.jpg" width="700px">
The relationship between final sequence reads of MiSeq-style and iSeq-style fastq files. Almost close to 1:1 line (dashed line), but the number of sequence reads of iSeq-style fastq files are slightly less than that of MiSeq-style fastq files.

## Figure 3: Relative abundance of MiSeq-style v.s. iSeq-style fastq files
<img src="06_ComparisonOut/RelativeAbundance_scatterplot.jpg" width="700px">
The relationship between relative abundances of MiSeq-style and iSeq-style fastq files. Almost on 1:1 line (dashed line), but the relative abundance of rare taxa of iSeq-style fastq files are slightly lower than that of MiSeq-style fastq files.


